base/compiler/typeinfer.jl

```diff
@@ -903,7 +905,7 @@ function typeinf_ext(interp::AbstractInterpreter, mi::MethodInstance)
             end
         end
     end
-    if ccall(:jl_get_module_infer, Cint, (Any,), method.module) == 0
+    if ccall(:jl_get_module_infer, Cint, (Any,), method.module) == 0 && !generating_sysimg()
         return retrieve_code_info(mi)
```

Comment:
Aren't you holding a lock here? This seems like a bad place for this `return` to happen. (and also shouldn't it just return `nothing`, to directly signal that inference failed)